### DISCUSSION TITLE
Advice on implementing custom validation with Prisma Client

### DISCUSSION THREAD

**Discussion by carolynwanf** [prisma/prisma#22058] [Q&A]
### Question

Hi! I'm new to Prisma and would really appreciate any advice you may have! I have a schema that I'm trying to implement [custom validation with Zod](https://www.prisma.io/docs/concepts/components/prisma-client/custom-validation) for, and this is my current plan:

1. Write Zod schemas for the models that need custom validation to validate email/url/other fields
2. Create a Prisma Client Extension that uses all of the Zod schemas I've written.

However, there are a couple points I'm stuck on:
1. Is there an easier / more scalable way that writing out all the Zod schemas by hand? I found this [library](https://www.npmjs.com/package/zod-prisma-types) for generating Zod schemas from Prisma models, but it doesn't seem to reliably generate schemas I want (ex: just generating z.string() instead of z.string().email())
2. I can't seem to figure out how to handle nested models, I keep getting errors like this:
`The types of '_type.emails' are incompatible between these types.
    Type '{ email: string; }[]' has no properties in common with type 'EmailInfoCreateNestedManyWithoutPersonInput'.`
even though this schema for the type of that field:
`const EmailInfoSchema = z.object({
  email: z.string().email(),
}) satisfies z.Schema<Prisma.EmailInfoCreateInput>;`
doesn't throw any errors.

For some context, here's a snippet of the schema I'm working on:
```prisma
/// Company represents a business entity.
model Company {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  // Fields that consistently exist across data sources.
  name   String /// The name of the company.
  domain String @unique /// The domain of the company.

  // Enrichments + additional information.
  rawData RawData[] /// Information about the Person that has been acquired via API calls, scraping, or other means. Information should be added here rather than the associated prospect companies, if it exists. This is like the customData field we currently use, but holds information about the source of the data and when it was created/last updated.

  /// Relations
  people          Person[] /// A list of people who work for this company.
  ProspectCompany ProspectCompany[]
}

/// Person represents an individual that works for a Company.
model Person {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  // Mandatory fields
  firstName       String /// The first name of the person.
  lastName        String /// The last name of the person.
  title           String? /// The person's job title.
  emails          EmailInfo[] /// The person's email. Each item keeps track of an email, the source of the email, and the last time bounce detection was run.
  linkedinProfile String      @unique /// The person's LinkedIn profile link. This should be a well-formatted URL. This is the only attribute that is *reliably* unique across people.

  // Enrichments + additional information.
  rawData RawData[] /// Information about the Person that has been acquired via API calls, scraping, or other means. Information should be added here rather than the associated prospect people, if it exists. This is like the customData field we currently use, but holds information about the source of the data and when it was created/last updated.

  companyId Int /// Foreign key to the associated company.
  company   Company @relation(fields: [companyId], references: [id])

  prospectPeople ProspectPerson[] /// A list of ProspectPeople that are related to the Person (meaning instances where the person is being targeted in a campaign).
}

/// EmailInfo represents detailed information about an individual's email, including the source and last verification date. All emails in the DB should be verified.
model EmailInfo {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @default(now())
  updatedAt DateTime @updatedAt

  email String /// The email itself.

  personId       Int? /// The associated person. This is optional because there might only be associated prospect person(s), which means the prospect is not uniquely identifiable by LinkedIn Profile.
  person         Person?          @relation(fields: [personId], references: [id])
  prospectPeople ProspectPerson[] /// The associated prospect person(s). There will probably only ever be one, but there might be more as our service scales.

  ApiCache   ApiCache? @relation(fields: [apiCacheId], references: [id])
  apiCacheId Int?      @unique /// Foreign key to the cache item that the `email` field comes from. The cache item should be a query to an email validator. If the associated cache item has expired, the data is stale. If the associated cache item is not populated, the email has yet to be validated.

  @@unique([email])
}
```
I would really appreciate and advice y'all might have! Please let me know if I can provide any additional context.

- OS: macOS
- Database: PostgreSQL
- Node.js version: v18.17.1

- Run `prisma -v` to see your Prisma version and paste it
```
prisma                  : 5.5.2
@prisma/client          : 5.6.0
```

**Comment by ludralph** [comment]
Hi @carolynwanf 

Thank you for raising this question and apologies for the late reply. This discussion slipped through the cracks.

> Is there an easier / more scalable way that writing out all the Zod schemas by hand? I found this [library](https://www.npmjs.com/package/zod-prisma-types) for generating Zod schemas from Prisma models, but it doesn't seem to reliably generate schemas I want (ex: just generating z.string() instead of z.string().email())

Have you tried out this library [zod-prisma](https://github.com/CarterGrimmeisen/zod-prisma)? This should hopefully fix the issue you are running into with generating Zod schemas from Prisma models.

> I can't seem to figure out how to handle nested models, I keep getting errors like this:
The types of '_type.emails' are incompatible between these types. Type '{ email: string; }[]' has no properties in common with type 'EmailInfoCreateNestedManyWithoutPersonInput'.
even though this schema for the type of that field:
const EmailInfoSchema = z.object({   email: z.string().email(), }) satisfies z.Schema<Prisma.EmailInfoCreateInput>;
doesn't throw any errors.

When exactly do you get these errors? Can you please share your query or a minimal reproduction so that we can investigate?

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
